metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala

```diff
@@ -0,0 +1,435 @@
+
+final class Ammonite(
+    buffers: Buffers,
+    compilers: Compilers,
+    compilations: Compilations,
+    statusBar: StatusBar,
+    diagnostics: Diagnostics,
+    doctor: Doctor,
+    tables: () => Tables,
+    languageClient: MetalsLanguageClient,
+    buildClient: ForwardingMetalsBuildClient,
+    userConfig: () => UserConfiguration,
+    indexWorkspace: () => Future[Unit],
+    workspace: () => AbsolutePath,
+    focusedDocument: () => Option[AbsolutePath],
+    buildTargets: BuildTargets,
+    buildTools: () => BuildTools
+)(implicit ec: ExecutionContextExecutorService)
+    extends Cancelable {
+
+  def buildServer: Option[BuildServerConnection] =
+    buildServer0
+  def lastImportedBuild: ImportedBuild =
+    lastImportedBuild0
+
+  private var buildServer0 = Option.empty[BuildServerConnection]
+  private var lastImportedBuild0 = ImportedBuild.empty
+  private var lastImportVersions = VersionsOption(None, None)
+
+  private val cancelables = new MutableCancelable()
+  private val isCancelled = new AtomicBoolean(false)
+  def cancel(): Unit = {
+    if (isCancelled.compareAndSet(false, true)) {
+      val buildShutdown = buildServer0 match {
+        case Some(build) => build.shutdown()
+        case None => Future.unit
+      }
+      try cancelables.cancel()
+      catch {
+        case NonFatal(_) =>
+      }
+      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
+      catch {
+        case _: TimeoutException =>
+      }
+    }
+  }
+
+  def loaded(path: AbsolutePath): Boolean =
+    path.isAmmoniteScript && {
+      val uri = path.toNIO.toUri.toASCIIString
+      lastImportedBuild0.targetUris.contains(uri)
+    }
+
+  private def connectToNewBuildServer(
+      build: BuildServerConnection
+  ): Future[BuildChange] = {
+    scribe.info(s"Connected to Ammonite Build server v${build.version}")
+    cancelables.add(build)
+    compilers.cancel()
+    buildServer0 = Some(build)
+    val importedBuild0 = // FIXME timed("imported ammonite build") {
+      MetalsLanguageServer.importedBuild(build)
+    for {
+      build0 <- statusBar.trackFuture(
+        "Importing Ammonite scripts",
+        importedBuild0
+      )
+      _ = {
+        lastImportedBuild0 = build0
+      }
+      _ <- indexWorkspace()
+      // _ = checkRunningBloopVersion(build.version)
+      _ <- Future.sequence[Unit, List](
+        compilations
+          .cascadeCompileFiles(buffers.open.toSeq) ::
+          compilers.load(
+            buffers.open.toSeq.filter(!_.isAmmoniteScript)
+          ) ::
+          Nil
+      )
+    } yield BuildChange.Reconnected
+  }
+
+  private def disconnectOldBuildServer(): Future[Unit] = {
+    if (buildServer0.isDefined)
+      scribe.info("disconnected: ammonite build server")
+    buildServer0 match {
+      case None => Future.unit
+      case Some(value) =>
+        // remove from cancellable too?
+        buildServer0 = None
+        lastImportedBuild0 = ImportedBuild.empty
+        cancelables.remove(value)
```

Comment:
Maybe we should just do `cancelables.cancel()` this will also remove the value.

Reply:
Yes, fixed.